extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/ReadMongoP.java

```diff
@@ -0,0 +1,482 @@
+            }
+            cursor = changeStream.batchSize(BATCH_SIZE).iterator();
+        }
+
+        @Override
+        boolean everCompletes() {
+            return false;
+        }
+
+        @Nonnull
+        @Override
+        public Traverser<?> nextChunkTraverser() {
+            try {
+                List<ChangeStreamDocument<Document>> chunk = readChunk();
+
+                Traverser<?> traverser = Traversers.traverseIterable(chunk)
+                                      .flatMap(doc -> {
+                                          resumeToken = doc.getResumeToken();
+                                          long eventTime = clusterTime(doc);
+                                          I item = mapFn.apply(doc);
+                                          return item == null
+                                                  ? Traversers.empty()
+                                                  : eventTimeMapper.flatMapEvent(item, 0, eventTime);
+                                      });
+                return traverser;
+            } catch (MongoException e) {
+                throw new JetException("error while reading from mongodb", e);
+            }
+        }
+
+        private List<ChangeStreamDocument<Document>> readChunk() {
+            List<ChangeStreamDocument<Document>> chunk = new ArrayList<>(BATCH_SIZE);
+            int count = 0;
+            boolean eagerEnd = false;
+            try {
+                while (count < BATCH_SIZE && !eagerEnd) {
+                    ChangeStreamDocument<Document> doc;
+                    doc = cursor.tryNext();
+                    if (doc != null) {
+                        chunk.add(doc);
+                        count++;
+                    } else {
+                        eagerEnd = true;
+                    }
+                }
+                return chunk;
+            } catch (MongoTimeoutException | MongoServerUnavailableException | MongoServerException e) {
+                logger.severe("Lost connection to MongoDB", e);
+                mongoClient = null;
+                return emptyList();
+            }
+        }
+
+        private long clusterTime(ChangeStreamDocument<Document> changeStreamDocument) {
+            BsonTimestamp clusterTime = changeStreamDocument.getClusterTime();
+            return clusterTime == null ? System.currentTimeMillis() : clusterTime.getValue();
+        }
+
+        @Nullable
+        @Override
+        public Object snapshot() {
+            return resumeToken;
+        }
+
+        @Override
+        public void restore(Object value) {
+            if (value != null) {
+                if (value instanceof BsonDocument) {
+                    this.resumeToken = (BsonDocument) value;
+                }
+            }
+        }
+
+        @Override
+        public void close() {
+            if (cursor != null) {
+                cursor.close();
+                cursor = null;
+            }
+        }
```

Comment:
Call `super.close()` to close the client instance?

Reply:
05b82e4c0fe8a88b002d1ee3c28afecd74eeac79